WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -136,19 +202,19 @@ private class WeeklyRoundupDataProvider {
     ///
     /// - Returns: the requested sites (could be an empty array if there's none) or an error if there is one.
     ///
-    private func getSites(result: @escaping (Result<[Blog], Error>) -> Void) {
+    private func getSites(in context: NSManagedObjectContext, result: @escaping (Result<[Blog], Error>) -> Void) {
 
-        switch getAllSites() {
+        switch getAllSites(in: context) {
         case .success(let sites):
-            filterCandidateSites(sites, result: result)
+            filterCandidateSites(sites, in: context, result: result)
         case .failure(let error):
             result(.failure(error))
```

Comment:
The `result` closure in the `success` case is called from the `context` queue, I think we should do the same in the `failure` case here, to keep the behaviour consistent:
```suggestion
            context.perform {
                result(.failure(error))
            }
```

Reply:
Good catch, I focused on the `success` case and forgot about the `failure` case.

I resolved the issue in 0659e51f8f9af011561d808846b7fd4b229fd7b9 but I took a different approach. Instead of wrapping each individual `result(...)` call in `context.perform`, I passed in a context-aware, thread-safe `result` callback. Like this:

```swift
self.getSites(in: context) { sitesResult in
   context.perform {
       result(sitesResult)
    }
}
```

This ensures the `result` callback is always called on the context's queue.

Reply:
I think this problem can be generalized like this:

Let's say we have an async function `foo`.

```swift
func foo(in context: NSManagedObjectContext, completion: () -> Void)
```

**The question is, who should be responsible for calling the `completion` on the context queue?**

And in my opinion, the answer depends on how many times `foo` function is called:

1. If `foo` is frequently called, it makes sense to incorporate `context.perform` directly within the `foo` function. This way, anyone using `foo` doesn't have to repeatedly call `context.perform` each time they use it.
2. But, if `foo` is a single-use function, like `getSites`, the responsibility of invoking `context.perform` can fall on the caller.

What do you think?

Reply:
I think the difficult part is not defining or implementing these conventions, rather it's the fact that we need to _remember_ these conventions. And, it's easy to miss code path where accidentally breaks these conventions.

Maybe we can take advantage of Swift's language feature to enforce these rules. Something like this:

```swift
protocol CallbackExecutor {
    func invokeCallback(_ closure: @escaping () -> Void)
}

extension DispatchQueue: CallbackExecutor {
    func invokeCallback(_ closure: @escaping () -> Void) {
        async {
            closure()
        }
    }
}

extension NSManagedObjectContext: CallbackExecutor {
    func invokeCallback(_ closure: @escaping () -> Void) {
        perform(closure)
    }
}

struct AsyncCallback<Arguments> {
    var executor: CallbackExecutor
    var closure: (Arguments) -> Void

    func invoke(with args: Arguments) {
        executor.invokeCallback {
            closure(args)
        }
    }

    static func queue<Arguments>(_ queue: DispatchQueue = .global(), closure: @escaping (Arguments) -> Void) -> AsyncCallback<Arguments> {
        .init(executor: queue, closure: closure)
    }

    static func context<Arguments>(_ context: NSManagedObjectContext, closure: @escaping (Arguments) -> Void) -> AsyncCallback<Arguments> {
        .init(executor: context, closure: closure)
    }
}

// Example:
func getSite(completion: AsyncCallback<Int>) {
    // ...
}

getSite(completion: .context(context, closure: { value in print(value) }))
```

Reply:
Resolving this conversation as this change is no longer needed. See this [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/21076#discussion_r1267526635).